react-app/src/components/SaveOutputForm.js

```diff
@@ -0,0 +1,213 @@
+  const [showNoFilesAlert, setShowNoFilesAlert] = useState(false);
+
+  function onSetOutputPath() {
+    window.api.getOutputPath().then((savePath) => {
+      if (!savePath.canceled) {
+        setOutputPath(savePath.filePaths[0]);
+      }
+    });
+  }
+
+  function clearOutputPath() {
+    setOutputPath('No Folder Selected');
+  }
+
+  function toggleSaveLogs() {
+    setSaveLogs(!saveLogs);
+  }
+
+  function getLabel(bundle, id) {
+    // attempt to get MRN and name with fhirpath
+    const patient = fhirpath.evaluate(bundle, "Bundle.descendants().resource.where(resourceType='Patient')")[0];
+    if (patient) {
+      const mrn = patient.id;
+      const name = patient.name[0].text;
+
+      const isMasked = fhirpath.evaluate(patient, 'Patient.identifier.extension.valueCode')[0] === 'masked';
+
+      // if both MRN and name -- return string w / both
+      if (typeof mrn === 'string' && mrn.length > 0 && typeof name === 'string' && name.length > 0 && !isMasked) {
+        const label = mrn.concat(': ').concat(name);
+        return label;
+      }
+      // if either MRN or name -- return string / the available one
+      if (typeof name === 'string' && name.length > 0) {
+        return name;
+      }
+      // if neither MRN nor name -- "Patient " + patient_resource_id
+      if (typeof mrn === 'string' && mrn.length > 0) {
+        return mrn;
+      }
+    }
+
+    // if no patient resource ID -- return "Patient " + props.id
+    let label = 'Patient';
+    label = label.concat(' ').concat(id.toString());
+    return label;
+  }
+
+  let defaultWhichFiles = {};
+  props.extractedData.forEach((bundle, i) => {
+    const label = getLabel(bundle, i);
+    defaultWhichFiles = {
+      ...defaultWhichFiles,
+      [label]: true,
+    };
+  });
+  const [whichFiles, setWhichFiles] = useState({ ...defaultWhichFiles });
+
+  // let whichFiles = {};
+
+  function togglePatientCheckbox(e) {
+    setWhichFiles({ ...whichFiles, [e.target.id]: e.target.checked });
+    if (!e.target.checked && selectAll) {
+      setSelectAll(false);
+    }
+  }
+
+  function getPatientCheckboxes() {
+    const list = props.extractedData.map((bundle, i) => {
+      const label = getLabel(bundle, i);
+      return (
+        <Form.Check
+          type="checkbox"
+          label={label}
+          key={label}
+          id={label}
+          checked={whichFiles[label]}
+          onChange={togglePatientCheckbox}
+        />
+      );
+    });
+    return list;
+  }
+
+  function toggleSelectAll() {
+    // modify which files will be saved first, so that there's no issues with state update delays
+    if (!selectAll) {
+      props.extractedData.forEach((bundle, i) => {
+        const label = getLabel(bundle, i);
+        defaultWhichFiles = {
+          ...defaultWhichFiles,
+          [label]: true,
+        };
+      });
+      setWhichFiles({ ...defaultWhichFiles });
+    }
+    // set select all
+    setSelectAll(!selectAll);
+  }
```

Comment:
I have made it so that when Select All is deselected, all of the file labels are unchecked. For the second part, do you mean that the value of the Select All checkbox will be set based on the values of the other checkboxes?

Reply:
I have made it so that Select All is checked when all of the other boxes are checked.

Reply:
Chiming in here – regarding Julia's second comment I think you understand her precisely Seneca. Instead of manually updating the value of SelectAll every time we update a patient-file checkbox/click the selectAll checkbox, we can define a function that tells us whether or not we've selected all the check boxes, and have that function's return value be used for the SelectAll checkbox. That would simplify the state management a bit

Reply:
I have moved the code to check if all files are selected into a separate function.

Reply:
I have removed the selectAll and setSelectAll variables, and replaced them with a function called isAllSelected, which checks if all the files are selected or not.